mne_bids/copyfiles.py

```diff
@@ -339,6 +340,94 @@ def copyfile_brainvision(vhdr_src, vhdr_dest, anonymize=None, verbose=False):
             print('Anonymized all dates in VHDR and VMRK.')
 
 
+def copyfile_edf(src, dest, anonymize=None):
+    """Copy an EDF, EDF+, or BDF file to a new location, optionally anonymizing
+    participant/patient information.
+
+    Parameters
+    ----------
+    src : str
+        The src path of the .edf or .bdf file to be copied.
+    dest : str
+        The destination path of the .edf or .bdf file.
+    anonymize : dict | None
+        If None (default), no anonymization is performed.
+        If dict, data will be anonymized depending on the keys provided with
+        the dict: `daysback` is a required key, `keep_his` is an optional key.
+
+        `daysback` : int
+            Number of days by which to move back the recording date in time.
+            In studies with multiple subjects the relative recording date
+            differences between subjects can be kept by using the same number
+            of `daysback` for all subject anonymizations. `daysback` should be
+            great enough to shift the date prior to 1925 to conform with BIDS
+            anonymization rules.
+
+        `keep_his` : bool
+            By default (False), all subject information next to the recording
+            date will be overwritten as well. If True, keep subject information
+            apart from the recording date.
+
+    See Also
+    --------
+    mne.io.anonymize_info
+
+    """
+    # Ensure source & destination extensions are the same
+    fname_src, ext_src = _parse_ext(src)
+    fname_dest, ext_dest = _parse_ext(dest)
+    if ext_src != ext_dest:
+        raise ValueError(f'Need to move data with same extension'
+                         f' but got "{ext_src}", "{ext_dest}"')
+
+    # Copy data prior to any anonymization
+    sh.copyfile(src, dest)
+
+    # Anonymize EDF/BDF data, if requested
+    if anonymize is not None:
+        raw = read_raw_edf(dest, preload=False, verbose=0)
+
+        # Get patient info, recording info, and recording date
+        with open(dest, 'rb') as f:
+            f.seek(8) # id_info field starts 8 bytes in
+            id_info = f.read(80).decode('ascii').rstrip()
+            rec_info = f.read(80).decode('ascii').rstrip()
+
+        # Parse metadata from file
+        if len(id_info) == 0 or len(id_info.split(' ')) != 4:
+            id_info = "X X X X"
+        if len(rec_info) == 0 or len(rec_info.split(' ')) != 5:
+            rec_info = "Startdate X X X X"
+        pid, sex, birthdate, name = id_info.split(' ')
+        startdate, admin_code, tech, equip = rec_info.split(' ')[1:5]
+
+        # Anonymize recording date, using 4-digit year from recording info
+        # if possible (regular EDF year only stores last two digits)
```

Comment:
with regular EDF you mean the "not-edf-plus"? ... and EDF+ and BDF *do* allow storing 4 digit year?

Would it be cleaner to read from the header whether it is edf, edf+, or bdf? ... not sure myself.

Reply:
I sort of phrased that poorly: basically, all forms of EDF and BDF support storing the year in a 4-digit format in the "recording info" region of the header. EDF+ is only special in that it requires those fields to be filled in using a certain format. However, that region of the header is frequently either left blank or filled in with "X"s for the fields, whereas the less-precise meas_date region is guaranteed to be there on all EDF/BDF files.

The ideal thing to do here would be to change `mne.io` so its EDF/BDF header parser tries to get the more precise start date from the recording_id region first and only falls back to the `meas_date` field if it's missing. That would ensure EDF and BDFs anonymized by MNE-BIDS would always load in the properly anonymized date.

Reply:
thanks for the clarification!

> The ideal thing to do here would be to change mne.io so its EDF/BDF header parser tries to get the more precise start date from the recording_id region first and only falls back to the meas_date field if it's missing

agreed that would be good

> That would ensure EDF and BDFs anonymized by MNE-BIDS would always load in the properly anonymized date.

however - only for those people who use MNE ... :) other tools (fieldtrip? EEGLAB?) may have other ways to load the dates